dc/s2n-quic-dc/events/connection.rs

```diff
@@ -1,26 +1,209 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
-#[event("application:write")]
-pub struct ApplicationWrite {
+#[event("stream:write_flushed")]
+pub struct StreamWriteFlushed {
     /// The number of bytes that the application tried to write
     #[measure("provided", Bytes)]
-    total_len: usize,
+    provided_len: usize,
 
     /// The amount that was written
     #[measure("committed", Bytes)]
     #[counter("committed.total", Bytes)]
-    write_len: usize,
+    committed_len: usize,
+
+    /// Indicates that the write was the final offset of the stream
+    is_fin: bool,
+
+    /// The amount of time it took to process the write request
+    ///
+    /// Note that this includes both any syscall and encryption overhead
+    #[measure("processing_duration", Duration)]
+    processing_duration: core::time::Duration,
+}
+
+#[event("stream:write_blocked")]
+pub struct StreamWriteBlocked {
+    /// The number of bytes that the application tried to write
+    #[measure("provided", Bytes)]
+    provided_len: usize,
+
+    /// Indicates that the write was the final offset of the stream
+    is_fin: bool,
+
+    /// The amount of time it took to process the write request
+    ///
+    /// Note that this includes both any syscall and encryption overhead
+    #[measure("processing_duration", Duration)]
+    processing_duration: core::time::Duration,
+}
+
+#[event("stream:write_errored")]
+pub struct StreamWriteErrored {
+    /// The number of bytes that the application tried to write
+    #[measure("provided", Bytes)]
+    provided_len: usize,
+
+    /// Indicates that the write was the final offset of the stream
+    is_fin: bool,
+
+    /// The amount of time it took to process the write request
+    ///
+    /// Note that this includes both any syscall and encryption overhead
+    #[measure("processing_duration", Duration)]
+    processing_duration: core::time::Duration,
+
+    /// The system `errno` from the returned error
+    errno: Option<i32>,
+}
+
+#[event("stream:write_shutdown")]
+pub struct StreamWriteShutdown {
+    /// The number of bytes in the send buffer at the time of shutdown
+    #[measure("buffer_len", Bytes)]
+    buffer_len: usize,
+
+    /// If the stream required a background task to drive the stream shutdown
+    #[bool_counter("background")]
+    background: bool,
+}
+
+#[event("stream:write_socket_flushed")]
+pub struct StreamWriteSocketFlushed {
+    /// The number of bytes that the stream tried to write to the socket
+    #[measure("provided", Bytes)]
+    provided_len: usize,
+
+    /// The amount that was written
+    #[measure("committed", Bytes)]
+    #[counter("committed.total", Bytes)]
+    committed_len: usize,
+
+    /// The amount of time it took to process the write request
+    ///
+    /// Note that this includes both any syscall and encryption overhead
+    #[measure("processing_duration", Duration)]
+    processing_duration: core::time::Duration,
```

Comment:
So I guess this implies we have at least 4 timestamps/write call? (start, start socket flush, end socket flush, end)? I'm a little worried that will get quite expensive quickly...

Reply:
yeah that's right. i was a bit concerned as well... do you think i should skip the socket ones for now?

Reply:
I removed it for now